### PR TITLE
Fix last paginated notification group only including data on a single notification

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Notification group details get filled with the range of fetched notifications, inclusive of first and last notifications of a group within a page.

However, when reaching the last page, this means a single notification is taken into account, even if the group contains more notification.

This PR changes it so that upon reaching the last page (fewer groups than expected are returned), the corresponding bound is ignored, and all notifications of the group in the paging direction are taken into account when backfilling the group information.

The code for this is a lot more awkward than I would like, in part due to not being able to use ActiveRecord scopes here.

### CODE REVIEW

File: app/controllers/api/v2/notifications_controller.rb [hunk 1]
```diff
@@ -80,7 +80,19 @@ def load_grouped_notifications
...
+
+      if @notifications.size < limit_param(DEFAULT_NOTIFICATIONS_LIMIT)
+        if params[:min_id].present?
+          pagination_range = @notifications.last.id...(params[:max_id]&.to_i)
+        else
+          range_start = params[:since_id]&.to_i
+          range_start += 1 unless range_start.nil?
+          pagination_range = range_start..(@notifications.first.id)
+        end
+      end
```
Comment (oneiros) [inline app/controllers/api/v2/notifications_controller.rb:93]: I found this to be really hard to follow. A little thing that might have helped me would have been to extract the conditions into helper methods that are named in a way that makes it clearer what cases these are.

So something like:

```ruby
if last_page?
  if paginating_down?
    # ...
  elsif paginating_up?
    # ...
  end
end
```

(I am not totally sure the naming makes sense, but I hope you get my meaning.)

**Comment by oneiros** [comment]
> This PR changes it so that upon reaching the last page (fewer groups than expected are returned)

Is there a similar concern when the last page contains exactly the number of groups that is expected per page? Or does the problem not happen then?

**Comment by ClearlyClaire** [comment]
> > This PR changes it so that upon reaching the last page (fewer groups than expected are returned)
> 
> Is there a similar concern when the last page contains exactly the number of groups that is expected per page? Or does the problem not happen then?

The concern is the same, but:
- having the same number of groups as requested is not enough to know we are on the last page
- it just means one extra query is needed to get the remaining notifications, as opposed to one query per notification

**Review by oneiros — APPROVED** [review]
